lib/membrane/rtp/session/report.ex

```diff
@@ -0,0 +1,137 @@
+defmodule Membrane.RTP.Session.Report do
+  @moduledoc false
+  alias Membrane.{RTCP, RTP, Time}
+  require Membrane.Logger
+
+  defmodule Data do
+    @moduledoc false
+    use Bunch.Access
+
+    @type t :: %__MODULE__{
+            remote_ssrcs: MapSet.t(RTP.ssrc_t()),
+            remote_reports: %{
+              RTP.ssrc_t() => %{
+                cut_wallclock_timestamp: pos_integer,
+                arrival_time: Time.t()
+              }
+            },
+            stats: [{RTP.ssrc_t(), RTP.ssrc_t(), RTP.JitterBuffer.Stats.t()}]
+          }
+    defstruct remote_ssrcs: MapSet.new(),
+              remote_reports: %{},
+              stats: []
+  end
+
+  @type maybe_report_t :: {:report, RTCP.CompoundPacket.t()} | :no_report
+
+  @spec init_report(ssrcs :: %{RTP.ssrc_t() => RTP.ssrc_t()}, Data.t()) ::
+          {maybe_report_t, MapSet.t(RTP.ssrc_t()), Data.t()}
+  def init_report(ssrcs, report_data) do
+    remote_ssrcs = ssrcs |> Map.keys() |> MapSet.new()
+
+    maybe_report =
+      if Enum.empty?(report_data.remote_ssrcs) do
+        :no_report
+      else
+        Membrane.Logger.warn(
+          "Not received stats from ssrcs: #{Enum.join(report_data.ssrcs, ", ")}"
+        )
+
+        {:report, generate_report(report_data)}
+      end
```

Comment:
Sending the report here doesn't seem like a good idea
- timeout for stats gathering should be separate from frequency of report generation
- if after this report the stats will be gathered quickly, we'd generate 2 reports one after another
- sending report from `init_report` is not quite intuitive

I would also consider changing the way reports are obtained: returned value would just contain atom indicating whether the report is ready and a separate function for obtaining it that would be used both when message is ready and when the timeout is reached

Reply:
> timeout for stats gathering should be separate from frequency of report generation

I believe that would only introduce unnecessary complexity. Stats should be always gathered quickly, and if they don't come till the next tick, there's a problem with the frequency, not the stats.

> if after this report the stats will be gathered quickly, we'd generate 2 reports one after another

that shouldn't happen, and even if it happens to, it doesn't hurt us

> I would also consider changing the way reports are obtained: returned value would just contain atom indicating whether the report is ready and a separate function for obtaining it that would be used both when message is ready and when the timeout is reached

Then we should support the case when one doesn't generate a report when it's ready. And I don't see any benefits of that.

Reply:
> introduce unnecessary complexity

The only unnecessary complexity I see is doing some weird stuff in `init_report` :man_shrugging: 

> Stats should be always gathered quickly, and if they don't come till the next tick, there's a problem with the frequency, not the stats

The keyword is "SHOULD". The assumption that it has to be a frequency problem is wrong. What if you have a bug causing JitterBuffer to get stuck? Or the message will not reach the target because of bug in some bin on the way? Or the bin's message queue is huge? Or the system is overloaded? Or maybe the SSRC was removed after sending a stats request? Besides, at some point we have to combine report sending frequency with the bandwidth and mixing that logic with stats gathering timeout is more of the unnecessary complexity.

>  it doesn't hurt us

It does break the bandwidth constraints from RFC. Of course, it's not a critical failure, however this is still a bug, shouldn't work this way and I don't see any benefit of not fixing it.

> when one doesn't generate a report when it's ready

What's here to support? Init should clear the state anyway and such scenario would occur if the bandwidth constraints would disallow sending the report.

Reply:
> The only unnecessary complexity I see is doing some weird stuff in init_report 🤷‍♂️

The other is adding useless timeouts and timers

> The keyword is "SHOULD" (...)

If we encounter any of these bugs, we will really have bigger problems than exceeding RTCP bandwidth constraints

Nevertheless, If you think we need an additional timeout for stats gathering, how should it be obtained/calculated?